File: desafios-python/desafio6/cpp/main.cpp
// ============================================================
// Desafio 6: Calculadora de Desconto
// Disciplina: Linguagem de Programação I - FATEC
// Descrição: Calcula o valor final de um produto após a
//            aplicação de uma porcentagem de desconto.
// ============================================================

#include <iostream>
#include <iomanip>
using namespace std;

// ----------------------------
// VERSÃO 2: Função Auxiliar
// (retorna -1.0 para sinalizar entrada inválida)
// ----------------------------
double aplicarDesconto(double valorProduto, double porcentagemDesconto) {
    if (porcentagemDesconto < 0 || porcentagemDesconto > 100)
        return -1.0;

    double valorDesconto = valorProduto * (porcentagemDesconto / 100.0);
    return valorProduto - valorDesconto;
}

// ----------------------------
// VERSÃO 1: Procedural
// ----------------------------
void versaoProcedural() {
    cout << "========================================" << endl;
    cout << "  VERSAO 1 - Procedural" << endl;
    cout << "========================================" << endl;

    double valorOriginal, descontoPct;
    cout << "Digite o valor original do produto (R$): ";
    cin >> valorOriginal;
    cout << "Digite a porcentagem de desconto (%): ";
    cin >> descontoPct;

    if (descontoPct < 0 || descontoPct > 100) {
        cout << "  X Porcentagem invalida. Deve ser entre 0 e 100." << endl;
        return;
    }

    double valorDesconto = valorOriginal * (descontoPct / 100.0);
    double valorFinal    = valorOriginal - valorDesconto;

    cout << fixed << setprecision(2);
    cout << "\n  Valor original:  R$ " << valorOriginal << endl;
    cout << "  Desconto (" << (int)descontoPct << "%): R$ " << valorDesconto << endl;
    cout << "  Valor final:     R$ " << valorFinal << endl;
}

// ----------------------------
// VERSÃO 2: Modular com Funções
// ----------------------------
void versaoModular() {
    cout << "========================================" << endl;
    cout << "  VERSAO 2 - Modular com Funcoes" << endl;
    cout << "========================================" << endl;

    double valorOriginal, descontoPct;
    cout << "Digite o valor original do produto (R$): ";
    cin >> valorOriginal;
    cout << "Digite a porcentagem de desconto (%): ";
    cin >> descontoPct;

    double valorFinal = aplicarDesconto(valorOriginal, descontoPct);

    if (valorFinal < 0) {
        cout << "  X Porcentagem invalida. Deve ser entre 0 e 100." << endl;
        return;
    }

    double valorDesconto = valorOriginal - valorFinal;

    cout << fixed << setprecision(2);
    cout << "\n  Valor original:  R$ " << valorOriginal << endl;
    cout << "  Desconto (" << (int)descontoPct << "%): R$ " << valorDesconto << endl;
    cout << "  Valor final:     R$ " << valorFinal << endl;
}

// ----------------------------
// Programa Principal
// ----------------------------
int main() {
    cout << "\n>>> DESAFIO 6: Calculadora de Desconto <<<\n" << endl;

    cout << "Qual versao deseja executar?" << endl;
    cout << "  1 - Procedural" << endl;
    cout << "  2 - Modular com Funcoes" << endl;
    cout << "Escolha (1 ou 2): ";

    int escolha;
    cin >> escolha;

    if (escolha == 1)
        versaoProcedural();
    else if (escolha == 2)
        versaoModular();
    else {
        cout << "Opcao invalida. Executando versao modular por padrao." << endl;
        versaoModular();
    }

    return 0;
}

File: desafios-python/desafio6/java/Main.java
// ============================================================
// Desafio 6: Calculadora de Desconto
// Disciplina: Linguagem de Programação I - FATEC
// Descrição: Calcula o valor final de um produto após a
//            aplicação de uma porcentagem de desconto.
// ============================================================

import java.util.Scanner;

public class Main {

    // ----------------------------
    // VERSÃO 2: Função Auxiliar
    // (retorna -1.0 para sinalizar entrada inválida)
    // ----------------------------
    public static double aplicarDesconto(double valorProduto, double porcentagemDesconto) {
        if (porcentagemDesconto < 0 || porcentagemDesconto > 100)
            return -1.0;

        double valorDesconto = valorProduto * (porcentagemDesconto / 100.0);
        return valorProduto - valorDesconto;
    }

    // ----------------------------
    // VERSÃO 1: Procedural
    // ----------------------------
    public static void versaoProcedural(Scanner sc) {
        System.out.println("========================================");
        System.out.println("  VERSAO 1 - Procedural");
        System.out.println("========================================");

        System.out.print("Digite o valor original do produto (R$): ");
        double valorOriginal = sc.nextDouble();
        System.out.print("Digite a porcentagem de desconto (%): ");
        double descontoPct = sc.nextDouble();

        if (descontoPct < 0 || descontoPct > 100) {
            System.out.println("  X Porcentagem invalida. Deve ser entre 0 e 100.");
            return;
        }

        double valorDesconto = valorOriginal * (descontoPct / 100.0);
        double valorFinal    = valorOriginal - valorDesconto;

        System.out.printf("%n  Valor original:  R$ %.2f%n", valorOriginal);
        System.out.printf("  Desconto (%.0f%%): R$ %.2f%n", descontoPct, valorDesconto);
        System.out.printf("  Valor final:     R$ %.2f%n", valorFinal);
    }

    // ----------------------------
    // VERSÃO 2: Modular com Funções
    // ----------------------------
    public static void versaoModular(Scanner sc) {
        System.out.println("========================================");
        System.out.println("  VERSAO 2 - Modular com Funcoes");
        System.out.println("========================================");

        System.out.print("Digite o valor original do produto (R$): ");
        double valorOriginal = sc.nextDouble();
        System.out.print("Digite a porcentagem de desconto (%): ");
        double descontoPct = sc.nextDouble();

        double valorFinal = aplicarDesconto(valorOriginal, descontoPct);

        if (valorFinal < 0) {
            System.out.println("  X Porcentagem invalida. Deve ser entre 0 e 100.");
            return;
        }

        double valorDesconto = valorOriginal - valorFinal;

        System.out.printf("%n  Valor original:  R$ %.2f%n", valorOriginal);
        System.out.printf("  Desconto (%.0f%%): R$ %.2f%n", descontoPct, valorDesconto);
        System.out.printf("  Valor final:     R$ %.2f%n", valorFinal);
    }

    // ----------------------------
    // Programa Principal
    // ----------------------------
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.println("\n>>> DESAFIO 6: Calculadora de Desconto <<<\n");
        System.out.println("Qual versao deseja executar?");
        System.out.println("  1 - Procedural");
        System.out.println("  2 - Modular com Funcoes");
        System.out.print("Escolha (1 ou 2): ");

        int escolha = sc.nextInt();

        if (escolha == 1)
            versaoProcedural(sc);
        else if (escolha == 2)
            versaoModular(sc);
        else {
            System.out.println("Opcao invalida. Executando versao modular por padrao.");
            versaoModular(sc);
        }

        sc.close();
    }
}
